### ISSUE TITLE
<simple_intropattern> as used in "apply ... in ... as <simple_intropattern>" not documented

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#9221]
#### Description of the problem

It appears that the documentation doesn't describe the syntax or semantics for <simple_intropattern> as used in `apply ... in ... as <simple_intropattern>` and several other tactics.

I will take a stab at writing this up.  Please advise me if it is indeed in the documentation somewhere.

**Comment by Zimmi48** [comment]
The documentation of intro patterns can be found here: https://coq.inria.fr/refman/proof-engine/tactics.html#coq:tacn.intros.
However, this documentation is very problematic as it is written now. In particular, it is not discoverable. This part should be augmented with a grammar. Only the tokens listed in this grammar should be used in the documentation of tactics. This would ensure that people can jump to the definitions by clicking on the tokens in the tactic description.

**Comment by jfehrle** [comment]
Maybe we can tweak the `.mlg` in simple ways, e.g. reordering entries, renaming nonterminals for clarity, etc.  But we may still have nonterminals that are shared between relatively unrelated topics.  But IIUC sphinx won't let you repeat a nonterminal in two different places, so only one topic would show the nonterminal and the others would just be linked--not so user friendly.

We could consider repeating the right hand side of an existing nonterminal definition under multiple new nonterminal names.  Could be a future maintenance issue if all the repeats should later need the same change.

On the other hand, changing the grammar would require careful scrutiny to ensure correctness.

From the WG discussion, I gather that the intro patterns are not very consistent but that improving that will take a long time (and some consensus on how to change them).  Makes it more fun to document.

Do you know if `or_and_intropattern` can be simplified to one line with a LIST0?  Who is a good person to ask about grammar changes (which are probably under many reviewers)?

**Comment by Zimmi48** [comment]
>But we may still have nonterminals that are shared between relatively unrelated topics

I don't get this. Do you have an example?

>IIUC sphinx won't let you repeat a nonterminal in two different places, so only one topic would show the nonterminal and the others would just be linked--not so user friendly.

AFAICT nonterminals can only be defined once in the parser and tokens can only be defined once in Sphinx. It seems consistent to me unless I'm missing something.

>On the other hand, changing the grammar would require careful scrutiny to ensure correctness.

I don't think there are much risks in trying to do that: CI and reviewers will catch mistakes.

>Do you know if or_and_intropattern can be simplified to one line with a LIST0? Who is a good person to ask about grammar changes (which are probably under many reviewers)?

I don't know much about parsers but @coq/parsing-maintainers should be able to help here :wink:

**Comment by jfehrle** [comment]
> I don't get this. Do you have an example?

`naming_intropattern` is used for intro patterns but also is referenced by `eqn_ipat` (used in the `remember` tactic). The latter is not part of the intro pattern topic.
```
  eqn_ipat:
    [ [ IDENT "eqn"; ":"; pat = naming_intropattern -> { Some (CAst.make ~loc pat) }
```
> AFAICT nonterminals can only be defined once in the parser and tokens can only be defined once in Sphinx. 

It appears that Sphinx lets you repeat definitions of nonterminals but AFAIK clicking on a reference to such a nonterminal can only go to one of the definitions (maybe always the first, I haven't checked).  Maybe useful even if imperfect.

**Comment by ejgallego** [comment]
My impression is that [unfortunately] definitions for the parser are a bit far from what you would like for documentation; likely some post-processing would be needed.

Playing with the post-processing should be possible tho; it shouldn't take long to expose the rules so SerAPI or other ML tool can process the rules [SerAPI for example can already dump grammar of notations]

**Comment by Zimmi48** [comment]
@jfehrle But this is still a nonterminal that's defined in a single place so it makes sense to define it only once in the manual. Tokens are clickable anyways, and thus readers can click to refer to the definition.

**Comment by jfehrle** [comment]
@Zimmi48 It's not that simple.  If I'm creating a `remember` tactic, do I need to first click on `eqn_epat` and then again on `naming_intropatter`?  And then back twice?  Readability is much better when everything is on the same page.  That's in tension with not defining things twice.  Humans are not well-equipped with dealing with multiple images at the same time.  But a judgement call, no hard rule.

**Comment by Zimmi48** [comment]
This `eqn: @naming_intropattern` feature exists in `remember`, `eremember` and `destruct`. There are two solutions to document it. Either
```
.. tacv:: destruct @ident eqn:@naming_intropattern
```
and the same with other tactics (in that case we don't define the `eqn_ipat` grammar token in the manual), or if we think it's worth factorizing, we do define it but then it would make sense to define it close to the grammar of intro-patterns. In any case, I strongly reject defining a grammar token twice in the manual, when it is only defined once in the parser. This can only lead to maintainability issues.
